consensus/src/dag/dag_store.rs

```diff
@@ -128,15 +114,31 @@ impl Dag {
     }
 
     pub fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        ensure!(
+            node.epoch() == self.epoch_state.epoch,
+            "different epoch {}, current {}",
+            node.epoch(),
+            self.epoch_state.epoch
+        );
         let node = Arc::new(node);
         let author = node.metadata().author();
         let index = *self
             .author_to_index
             .get(author)
             .ok_or_else(|| anyhow!("unknown author"))?;
         let round = node.metadata().round();
-        ensure!(round >= self.lowest_round(), "round too low");
-        ensure!(round <= self.highest_round() + 1, "round too high");
+        ensure!(
+            round >= self.lowest_round(),
+            "round too low {}, lowest in dag {}",
+            round,
+            self.lowest_round()
+        );
+        ensure!(
+            round <= self.highest_round() + 1,
+            "round too high {}, highest in dag {}",
+            round,
+            self.highest_round()
+        );
```

Comment:
re: line 154 " self.storage.save_certified_node(&node)?;". We can save this storage call when add_node is called from Dag:new()

Reply:
good point